MAVProxy/modules/mavproxy_paraedit/param_editor.py

```diff
@@ -0,0 +1,212 @@
+
+            self.event_queue_lock.release()
+
+            # if event processing operations require a mission referesh in GUI
+            # (e.g., after a load or a verified-completed write):
+            if (request_read_after_processing_queue):
+                self.event_queue_lock.acquire()
+                self.event_queue.put(ParamEditorEvent(ph_event.PEE_READ_PARAM))
+                self.event_queue_lock.release()
+
+            # periodically re-request WPs that were never received:
+            # DON'T NEED TO! -- wp module already doing this
+
+            time.sleep(0.2)
+
+
+class ParamEditorMain(object):
+    def __init__(self, mpstate):
+        self.param_received = {}
+
+        self.event_queue = multiproc.Queue()
+        self.event_queue_lock = multiproc.Lock()
+        self.gui_event_queue = multiproc.Queue()
+        self.gui_event_queue_lock = multiproc.Lock()
+
+        self.close_window = multiproc.Semaphore()
+        self.close_window.acquire()
+
+        self.child = multiproc.Process(
+                        target=self.child_task,
+                        args=(self.event_queue,
+                              self.event_queue_lock, self.gui_event_queue,
+                              self.gui_event_queue_lock, self.close_window))
+        self.child.start()
+
+        self.event_thread = ParamEditorEventThread(
+                            self, self.event_queue, self.event_queue_lock)
+        self.event_thread.start()
+
+        self.mpstate = mpstate
+        self.mpstate.param_editor = self
+
+        self.last_unload_check_time = time.time()
+        self.unload_check_interval = 0.1  # seconds
+
+        self.time_to_quit = False
+        self.mavlink_message_queue = multiproc.Queue()
+        self.mavlink_message_queue_handler = threading.Thread(
+            target=self.mavlink_message_queue_handler)
+        self.mavlink_message_queue_handler.start()
+
+    def mavlink_message_queue_handler(self):
+        while not self.time_to_quit:
+            while True:
+                if not self.mavlink_message_queue.empty():
+                    break
+                time.sleep(0.1)
+            m = self.mavlink_message_queue.get()
+
+            # MAKE SURE YOU RELEASE THIS LOCK BEFORE LEAVING THIS METHOD!!!
+            # No "return" statement should be put in this method!
+            self.gui_event_queue_lock.acquire()
+
+            try:
+                self.process_mavlink_packet(m)
+            except Exception as e:
+                print("Caught exception (%s)" % str(e))
+                import traceback
+                traceback.print_stack()
+
+            self.gui_event_queue_lock.release()
+
+    def unload(self):
+        '''unload module'''
+        self.mpstate.param_editor.close()
+        self.mpstate.param_editor = None
+
+    def idle_task(self):
+        now = time.time()
+        if self.last_unload_check_time + self.unload_check_interval < now:
```

Comment:
As a general rule you want to subtract times whereever possible rather than adding intervals to a time.  Doesn't matter so much in Python, but does elsewhere.